Hqq serialization

mobicham opened the pull request:
Follow-up to https://github.com/huggingface/transformers/pull/32379
The goal of this PR is to add full support to save/load HQQ-quantized models directly in transformers. So far, serialization was done on the hqq-lib side via the `.pt` format which is not safe and doesn't work with very large models (>100B params) since the model is not sharded.

What was done during this PR:
- Make sure saving/loading HQQ-quantized models works properly.
- Make sure multi-gpu support works with the hqq backends (required some updates the hqq lib side)
- Make sure adding biases in architectures that do not have a bias by default works (biases are used in some HQQ-calibrated models)
- Added `update_expected_keys()` call in the quantizer. This allows loading quantized models that were initialized with `torch.nn.Linear` instead

Full gist to try it out: https://gist.github.com/mobicham/701dd564c52590203ee09631425ad797

mobicham committed:
HQQ model serialization attempt

mobicham committed:
Merge branch 'huggingface:main' into main

SunMarc committed:
fix hqq dispatch and unexpected keys

SunMarc committed:
Merge remote-tracking branch 'upstream/main' into hqq_serialization

SunMarc committed:
Merge remote-tracking branch 'upstream/main' into hqq_serialization

SunMarc committed:
style

mobicham committed:
remove check_old_param

mobicham commented:
https://github.com/huggingface/transformers/pull/33141/commits/5cb7d81547908dea660f525be5f77d9065b6edeb 
Removed the check_old_param hack. 
The problem however is that `HQQLinear` state_dict is huge, which makes loading extremely slow. So I added `run_expected_keys_check` which skips those checks for HQQLinear params. I am not sure if it's a clean way. If you just init a dummy `HQQLinear` you wouldn't get all the state_dict params anyway :thinking:  , let me know if there's a better way of doing this

SunMarc commented on src/transformers/quantizers/quantizer_hqq.py:
```diff
@@ -122,21 +139,47 @@ def create_quantized_param(
+                    unexpected_keys.remove(k)
+
+        if self.pre_quantized:
+            if isinstance(module, (torch.nn.Linear, HQQLinear)):
+                hqq_layer = HQQLinear(
+                    linear_layer=None,
+                    quant_config=None,
+                    compute_dtype=self.torch_dtype,
+                    device=target_device,
+                )
+
+            hqq_layer.axis = None
+            hqq_layer.channel_wise = None
+            hqq_layer.load_state_dict(module_state_dict)
```
In my PR, I have two cases: 
- if module is HQQLinear, we do a return 
- if module is a Linear, we replace it by hqqlinear + load_state dict 

Here, we do the same. This will make the loading way slower since we load the layer multiple times and it is probably what you are experiencing.

mobicham replied:
Reverting back to that version :+1:

SunMarc replied:
Maybe sure to push the commit ! To me, it's still not reverted.

mobicham replied:
it;s here
https://github.com/huggingface/transformers/blob/cbe219fe3ab5694f1fd2ace6597cb74bd8324d9a/src/transformers/quantizers/quantizer_hqq.py#L154-L178

mobicham replied:
Apologies for the confusion, it actually works with the reverted changes: https://github.com/huggingface/transformers/pull/33141/commits/2bb974c5e4faae8e110fa3bb42f22b0a1e5f7736

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -857,9 +858,14 @@ def _load_state_dict_into_meta_model(
 
     is_torch_e4m3fn_available = hasattr(torch, "float8_e4m3fn")
 
+    # We add this because HQQLinear dict has a very large state_dict (19 params/per module), which makes loading extremely slow
+    run_expected_keys_check = True
+    if isinstance(hf_quantizer, HqqHfQuantizer):
+        run_expected_keys_check = False
+
```
with the comment below, this might not be needed

mobicham replied:
Unfortunately we still need that check otherwise it breaks 
```Python
File ~/py310_env2/lib/python3.10/site-packages/accelerate/utils/modeling.py:358, in set_module_tensor_to_device(module, tensor_name, device, value, dtype, fp1
6_statistics, tied_params_map)
    355     return
    357 if old_value.device == torch.device("meta") and device not in ["meta", torch.device("meta")] and value is None:
--> 358     raise ValueError(f"{tensor_name} is on the meta device, we need a `value` to put in on {device}.")
    360 if value is not None:
    361     if old_value.shape != value.shape:

ValueError: weight is on the meta device, we need a `value` to put in on cuda:0
```

SunMarc replied:
Could you explain what happens here ? I don't understand why we are getting this error if we don't set that

mobicham replied:
Answer here: https://github.com/huggingface/transformers/pull/33141/#issuecomment-2314984385

SunMarc replied:
Can you change the comment as it is not what's happening ?

mobicham replied:
Resolved, see https://github.com/huggingface/transformers/pull/33141#issuecomment-2317474476 for the new version

SunMarc reviewed (commented):
Nice ! Let's fix the issue regarding the torchao backend and we can merge this. I left a few comments

mobicham commented:
2/3: Multi-gpu loading
Loading on multi-gpu looks like it's working fine. There's an issue with the BitBlas backend I just reported <a href="https://github.com/microsoft/BitBLAS/issues/154">here</a>
Forcing the input to use the same device was done on the <a href="https://github.com/mobiusml/hqq/commit/1e620a91c3453c7655925cf9c39d1beed14d518e">hqq</a> lib side.

3/3: state_dict on the same safetensor chunk
I run tests with different models and it's working fine (<a href="https://gist.github.com/mobicham/5c469f6c9e8244fb31fef99e1d5ccb13"> gist</a>):
```Python
model_id  = 'meta-llama/Meta-Llama-3-8B-Instruct' #OK
model_id  = 'meta-llama/Meta-Llama-3-70B' #OK 
model_id = "facebook/opt-125m" #OK
model_id = "meta-llama/Llama-2-13b-chat-hf" #OK
model_id = "microsoft/Phi-3-mini-128k-instruct" #OK
model_id = "google/gemma-2-9b-it" #OK
model_id = "google/gemma-2-2b" #OK
```

 so I think for the moment we can leave it until someone reports some issue,  I can't reproduce the problem anyway.


Next steps:
-  Revisit the comments above (@mobicham )
- Change/disable settings for hqqConfig because now saving/loading doesn't support quant scales/zeros as well as meta-data offloading. Need to deprecate it as well on the hqq lib side and a new pip version 2.0.0 (@mobicham )

mobicham committed:
revert to check HQQLinear in quantizer_hqq.py

mobicham committed:
revert to check HQQLinear in quantizer_hqq.py

mobicham committed:
update HqqConfig default params

mobicham commented:
@SunMarc 
- Reverted back to `if isinstance(module, (torch.nn.Linear, HQQLinear)):` but we still need that  `run_expected_keys_check` otherwise it breaks
- Updated the default `HqqConfig` default params since `quant_scale`, `quant_zero`, and `offload_meta` are now deprecated. Also done on the hqq-lib side. I also updated the tests, the doc and made a new hqq lib pip release `0.2.0`

mobicham committed:
make ci happy

SunMarc commented on tests/quantization/hqq/test_hqq.py:
```diff
@@ -146,22 +143,43 @@ def test_f16_quantized_model_with_offloading(self):
     def tearDown(self):
         cleanup()
 
-    def test_fp16_quantized_model_multipgpu(self):
+    def test_model_serialization(self):
         """
-        Simple LLM model testing fp16 with multi-gpu
+        Simple HQQ LLM save/load test
         """
-
-        quant_config = HqqConfig(nbits=8, group_size=64, quant_zero=False, quant_scale=False, axis=0)
+        quant_config = HqqConfig(nbits=4, group_size=64)
 
         hqq_runner = HQQLLMRunner(
-            model_id=MODEL_ID, quant_config=quant_config, compute_dtype=torch.float16, device="auto"
+            model_id=MODEL_ID, quant_config=quant_config, compute_dtype=torch.float16, device=torch_device
         )
 
-        check_hqqlayer(self, hqq_runner.model.model.layers[0].self_attn.v_proj)
-        check_forward(self, hqq_runner.model)
```
it would be nice to keep the multi-gpu test

mobicham replied:
Oh I think I removed it by mistake, let me take a look at it

mobicham replied:
It was not removed actually, it's still there: https://github.com/huggingface/transformers/blob/cbe219fe3ab5694f1fd2ace6597cb74bd8324d9a/tests/quantization/hqq/test_hqq.py#L121-L141

SunMarc replied:
oh thx for checking !

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -215,11 +209,8 @@ def __init__(
         self,
         nbits: int = 4,
         group_size: int = 64,
-        quant_zero: bool = True,
-        quant_scale: bool = False,
-        offload_meta: bool = False,
```
Are these deprecated in the latest version of hqq ? Would it make sense to enforce the user to download the latest version of hqq ?

mobicham replied:
Yeah, I saw that transformers installs the latest version of hqq in the docker file

SunMarc replied:
Then, maybe we can update `is_hqq_available` to force the user to install the latest hqq.

mobicham replied:
Like this https://github.com/huggingface/transformers/pull/33141/commits/9f7c23577b2f13423250dcb766dec0d421fe71ea ?

SunMarc replied:
yes could you also update the `validate_environment` function where this check is used to remind the user that he needs to upgrade their version ?

SunMarc replied:
>  Also, we can potentially let the user know that these args are deprecated if they set it in the latest version of hqq.

if we see these in the kwargs, we need to warn them that they are deprecated.

mobicham replied:
`validate_environment` message mentioning at least `0.2.0`: https://github.com/huggingface/transformers/pull/33141/commits/4682a725c52ab2b2bb00e021b6e145d1f6d0abb6

Deprecated keys in kwargs message:
https://github.com/huggingface/transformers/pull/33141/commits/813ed627b3079f53e0423f0f24a3be78a30c169c

SunMarc replied:
nice !

mobicham committed:
make ci happy

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -215,11 +209,8 @@ def __init__(
-        quant_scale: bool = False,
-        offload_meta: bool = False,
         view_as_float: bool = False,
-        axis: int = 0,
```
Breaking change here. Is this the recommended value now ?

mobicham replied:
Yes, I have been planning to do it for a long time, because only `axis=1` supports the fast backends like `torchao` and `bitblas`, so the new default value is `axis=1` instead of `axis=0`

SunMarc replied:
Well, I will leave it to you but I think it will be great to let the user know that axis have been set to 1 since it supports fast backends. How I would do is to set axis = None and in the post_init, if axis is None, we set it to 1 + add a `logger.info`

mobicham replied:
`axis` is required and it should be either 0 or 1 actually. It's just that before it was 0 and now it's 1.
What should the `logger.info` say?

mobicham replied:
I think the diff doesn't show exactly what's going on: https://github.com/huggingface/transformers/blob/ff982b3b2c8e9f4580ad0a32b90c40b3dcae5985/src/transformers/utils/quantization_config.py#L186-L217

SunMarc replied:
see new review

SunMarc reviewed (commented):
Added a couple of comments !

mobicham committed:
revert to HQQLinear check in quantizer_hqq.py

mobicham commented:
Regarding this: https://github.com/huggingface/transformers/pull/33141#discussion_r1734388659
The issue is that to remove that additional check, we need to have all the HQQLinear dict keys for each layer in the list of expected keys. There are 19 keys  per HQQLinear module. For a small model like `LLama3-8B`, that means `32*7*19=4256` checks per parameter which is extremely slow

mobicham committed:
check hqq_min version 0.2.0

mobicham committed:
Merge branch 'main' into hqq_serialization

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -216,11 +210,8 @@ def __init__(
-        offload_meta: bool = False,
         view_as_float: bool = False,
-        axis: int = 0,
+        axis: int = 1,
```
```suggestion
        axis: Optional[int] = None,
```

Then add in the init : 
```py
if axis is None: 
   axis=1
   logger.info("Setting axis=1 as faster backends such as torchao or bitblas are only compatible with it.")
```

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -193,15 +193,9 @@ class HqqConfig(QuantizationConfigMixin):
         view_as_float (`bool`, *optional*, defaults to `False`):
             View the quantized weight as float (used in distributed training) if set to `True`.
-        axis (`int`, *optional*, defaults to 0):
+        axis (`int`, *optional*, defaults to 1):
```
```suggestion
        axis (`Optional[int]`, *optional*, defaults to `None`):
```

SunMarc reviewed (commented):
Left a suggestion about axis

mobicham committed:
set axis=1 as default in quantization_config.py

mobicham committed:
Merge branch 'hqq_serialization' of https://github.com/mobiusml/transformers into hqq_serialization

mobicham committed:
validate_env with hqq>=0.2.0 version message

mobicham committed:
deprecated hqq kwargs message

mobicham committed:
make ci happy

SunMarc reviewed (approved):
Thanks for iterating ! Excited to see HQQ models on the hub 🔥

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33141). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

mobicham commented:
There's 2 TODOs to be done before merging:
- Check if adding a bias on architectures that don't support bias by default breaks the hqq model loading
- Trying to get rid of `run_expected_keys_check` by updating the `expected_keys`. This will require some modification on the hqq lib side as well to return the list of all the valid keys of the state dict

mobicham committed:
remove run_expected_keys_check hack + bump to 0.2.1 min hqq version

mobicham commented:
> There are TODOs to be done before merging:
> 
> * Check if adding a bias on architectures that don't support the bias by default breaks the hqq model loading.

✅  Checked, all good!

> * Trying to get rid of `run_expected_keys_check` by updating the `expected_keys`. This will require some modification on the hqq lib side as well to return the list of all the valid keys of the state dict. Then bump up min hqq lib version in transformers.

✅  Removed the `run_expected_keys_check` hack by extending `expected_keys` in the case where `HQQLinear` state dict is loaded but `Linear` is present instead.

SunMarc reviewed (commented):
Nice update ! left a few comments

mobicham commented on src/transformers/modeling_utils.py:
```diff
@@ -4104,6 +4112,41 @@ def _load_pretrained_model(
         expected_keys = list(model_state_dict.keys())
```
I didn't touch that line, it's from the master branch: https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_utils.py#L4104

SunMarc replied:
Yes, but i thought that the loading would be slow because of the big expected_keys list. Is it not the case anymore ?

mobicham replied:
I use sets inside the update function, it looks like it's working  just fine :thinking:

SunMarc replied:
Nice, works for me !

mobicham committed:
fix unexpected_keys hqq update

mobicham commented on src/transformers/modeling_utils.py:
```diff
@@ -4104,6 +4112,41 @@ def _load_pretrained_model(
+
+        if isinstance(hf_quantizer, HqqHfQuantizer):
+            expected_keys = _fix_hqq_keys(expected_keys)
+
```
https://gist.github.com/mobicham/701dd564c52590203ee09631425ad797

The check takes care of both cases. The problem is that, `pre_quantized` will return True even when the model is not pre-quantized because the model config already contains the quantization_config at the loading time. But it's fine, the update function works as expected.

You can test it with this gist: https://gist.github.com/mobicham/701dd564c52590203ee09631425ad797

SunMarc commented on src/transformers/quantizers/quantizer_hqq.py:
```diff
@@ -99,6 +99,52 @@ def update_missing_keys(
 
+    # Adds missing keys for HQQLinear modules that are loaded but the model with initialized with torch.nn.Linear
+    def update_expected_keys(self, model: "PreTrainedModel", keys: List[str], loaded_keys: List[str]) -> List[str]:
+        # Collects all quantizable (linear) layers
+        def _find_hqq_quantizable_layers(model, layers):
+            for name, module in model.named_children():
+                if isinstance(module, (torch.nn.Linear)):
+                    layers.add(module.name)
+                _find_hqq_quantizable_layers(module, layers)
+
+        new_keys = set(keys)
+        if is_hqq_available():
+            from hqq.core.quantize import HQQLinear
```
add a `if self.pre_quantized` condition as we shouldn't modify the keys when the model is not quantized yet.

mobicham replied:
Sure, works with the check! https://github.com/huggingface/transformers/pull/33141/commits/9053ad5f4c4bea7e396f12896d072c0999dda74c

mobicham committed:
add pre_quantized check

mobicham committed:
add update_expected_keys to base quantizerr

mobicham committed:
ci base.py fix?

mobicham committed:
ci base.py fix?

mobicham commented:
You can test with this gist: https://gist.github.com/mobicham/701dd564c52590203ee09631425ad797

SunMarc commented:
Nice could you just update a bit the description of the PR ?

mobicham commented:
@ArthurZucker just a friendly reminder to review this PR when you have a moment. Let me know if you need any clarifications or if there’s anything I can help with. Thank you very much :pray:

blap commented:
Waiting for this!

ArthurZucker commented on src/transformers/modeling_utils.py:
```diff
@@ -917,12 +917,17 @@ def _load_state_dict_into_meta_model(
             old_param = getattr(old_param, split)
+            # Not all the attributes of a module are Parameters/Tensor
+            if not isinstance(old_param, (torch.nn.Parameter, torch.Tensor)):
+                old_param = None
```
Is this for potentially ints ?

mobicham replied:
Yes, some parameters are strings (packing format, etc.), booleans or integers. They are necessary meta-data to dequantize

ArthurZucker commented on src/transformers/utils/quantization_config.py:
```diff
@@ -216,18 +210,23 @@ def __init__(
 
+        for deprecated_key in ["quant_zero", "quant_scale", "offload_meta"]:
+            if deprecated_key in kwargs:
+                logger.info(deprecated_key + " is deprecated. This parameter will be ignored in quantzation settings.")
```
```suggestion
                logger.info(deprecated_key + " is deprecated. This parameter will be ignored in quantization settings.")
```

ArthurZucker commented on src/transformers/quantizers/quantizer_hqq.py:
```diff
@@ -91,6 +91,65 @@ def validate_environment(self, *args, **kwargs):
+        else:
+            return missing_keys
+
+    # Adds missing keys for HQQLinear modules that are loaded but the model with initialized with torch.nn.Linear
+    def update_expected_keys(
```
I am not sure I understand exactly why we need to do this post loading vs while loading + is gonna be quite annoying to maintain as it feels like there's a lot of hacks but fine for me!

mobicham replied:
Unfortunately, this was the only way to make this work. The issue is that, when the model is created, it creates `nn.Linear` layers that have a "weight" parameter, but there's no "weight" parameter in `HQQLinear`, so when it tries to load the `HQQLinear` state-dict, it breaks.  I agree it's a bit complicated, actually the whole `HQQLinear` support has been complicated.

I think one better way of doing this in the future is to have a native `QuantLinear` layer officially supported by HF transformers, which could potentially integrate any quant method. I pitched this idea to Marc as well, I think it's worth thinking about. Happy to assist with that as well.

ArthurZucker reviewed (approved):
Thanks and sorry for the late review, just some small nits

blap commented:
Just for curiosity, what miss to merge?

SunMarc commented:
> Just for curiosity, what miss to merge?

Waiting for @mobicham to check the latest review and give me to heads-up to merge ! This should be done soon !

mobicham committed:
fix "quantization typo" src/transformers/utils/quantization_config.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

mobicham committed:
Merge branch 'main' into hqq_serialization

mobicham committed:
fix post merge

mobicham commented:
There's some issue with Llama3 specifically that just popped up after the merge, trying to figure out what's going on

SunMarc commented:
Thanks for iterating @mobicham! Merging!

rohit-gupta commented:
@mobicham minor documentation issue, but the transformers documentation page for quantization has a giant features matrix which still says serialization of HQQ models is not supported


https://huggingface.co/docs/transformers/main/quantization/overview

SunMarc commented:
Would you like to open a PR to fix this @rohit-gupta ?

mobicham commented:
@rohit-gupta thanks for flagging !

blap commented:
now model.save_pretrained(save_path) give this:



```

Traceback (most recent call last):
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\hqq1b.py", line 35, in <module>
    model.save_pretrained(save_path)
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\transformers\modeling_utils.py", line 2932, in save_pretrained
    state_dict_split = split_torch_state_dict_into_shards(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_torch.py", line 330, in split_torch_state_dict_into_shards
    return split_state_dict_into_shards_factory(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_base.py", line 108, in split_state_dict_into_shards_factory
    storage_id = get_storage_id(tensor)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_torch.py", line 382, in get_torch_storage_id
    if tensor.device.type == "meta":
       ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'device'


```

mobicham commented:
@blap is this related to the latest transformer changes? Otherwise, which hqq version causes this?

blap commented:
> @blap is this related to the latest transformer changes? Otherwise, which hqq version causes this?

I think so. I didn't had this problem in the release of hqq in transformers.
hqq version: 0.2.3

mobicham commented:
> > @blap is this related to the latest transformer changes? Otherwise, which hqq version causes this?
> 
> I think so. I didn't had this problem in the release of hqq in transformers. hqq version: 0.2.3 transformers version: 4.47.0.dev0

@SunMarc do you know what was changed by any  chance?

blap commented:
Transformers version 4.48.0.dev0 still has this problem...

mobicham commented:
Any one from the HF team can track down this problem please? What changed ? Nothing on the hqq lib side changed much.

blap commented:
@SunMarc ?

SunMarc commented:
Can you share your script @blap ? I'll have a look asap !

blap commented:
> Can you share your script @blap ? I'll have a look asap !




```

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, HqqConfig

model_id      = "mllmTeam/PhoneLM-1.5B"
repo          = "PhoneLM-1.5B"
nbits         = 4
group_size    = None
axis          = 0
save_path     = repo+"-nbits"+str(nbits)+"-GS"+str(group_size)+"-Axis"+str(axis)+"-HQQ2"
cache_dir     = repo+"-cache"
device        = "cpu"
compute_dtype = torch.float16

#Quantize
quant_config  = HqqConfig(nbits=nbits, group_size=group_size, axis=axis, quant_scale=False, quant_zero=False)

#Load the model
model = AutoModelForCausalLM.from_pretrained(
    model_id, 
    torch_dtype=compute_dtype, 
    cache_dir=cache_dir,
    device_map=device, 
    quantization_config=quant_config,
    low_cpu_mem_usage=True,
    trust_remote_code=True
)

#Tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_id, cache_dir=cache_dir)

# Save
model.save_pretrained(save_path)
tokenizer.save_pretrained(save_path)

```

Error:




```

Traceback (most recent call last):
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\hqq1b.py", line 32, in <module>
    model.save_pretrained(save_path)
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\transformers\modeling_utils.py", line 2971, in save_pretrained
    state_dict_split = split_torch_state_dict_into_shards(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_torch.py", line 369, in split_torch_state_dict_into_shards
    return split_state_dict_into_shards_factory(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_base.py", line 108, in split_state_dict_into_shards_factory
    storage_id = get_storage_id(tensor)
                 ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\huggingface_hub\serialization\_torch.py", line 746, in get_torch_storage_id
    if tensor.device.type == "meta":
       ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'device'


```

blap commented:
So...
Any ideas how to save?

mobicham commented:
@blap why don't you use the latest release ? It works fine last time I tried (last week)

blap commented:
> @blap why don't you use the latest release ? It works fine last time I tried (last week)

Which version do you use?

Version 4.45.0 give me this:


```

Traceback (most recent call last):
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\hqq1b.py", line 37, in <module>
    model.save_pretrained(save_path)
  File "C:\Users\Admin\Desktop\Python\0.LLMs\hqq\venv\Lib\site-packages\transformers\modeling_utils.py", line 2565, in save_pretrained
    raise ValueError(
ValueError: The model is quantized with QuantizationMethod.HQQ and is not serializable - check out the warnings from the logger on the traceback to understand the reason why the quantized model is not serializable.


```

mobicham commented:
@blap `4.47.0` works for sure

blap commented:
> @blap `4.47.0` works for sure

I just got the same error in this version too.
I tried others models without success.
I only run properly only on hqq, not on transformers.
Can you show some code, please?

mobicham commented:
@blap 
```Python
# pip install transformers==4.47.0;
# pip install hqq --upgrade;
##################################################################
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, HqqConfig

model_path = "meta-llama/Meta-Llama-3-8B-Instruct"
quant_model = "quant_model"

quant_config = HqqConfig(nbits=4, group_size=64, axis=1)

model = AutoModelForCausalLM.from_pretrained(model_path,
                                            torch_dtype=torch.float16,
                                            cache_dir='.',
                                            device_map="cuda:0",
                                            quantization_config=quant_config,
                                            low_cpu_mem_usage=True)

tokenizer = AutoTokenizer.from_pretrained(model_path)

model.save_pretrained(quant_model)
tokenizer.save_pretrained(quant_model)
```

blap commented:
I found the problem:
If I use group_size=None I got the error.

mobicham commented:
> I found the problem: If I use group_size=None I got the error.
Hmm interesting, will take a look into it, thanks for flagging it !
